File: validation/surface_flow_cases.jl
using Arrhenius
using NPZ
using Test
using TOML

include(joinpath(@__DIR__,"..","example","reactors","surf_pfr.jl"))
include(joinpath(@__DIR__,"..","example","reactors","surf_pfr_chain.jl"))

function validate_surface_flows(directory;run=true)
    path = joinpath(directory,"pox.surface.npz")
    flow = methane_surface_pfr(path)
    if run
        direct = surf_pfr(path;output=joinpath(directory,"flow.native.csv"))
        npzwrite(joinpath(directory,"flow.native.npz"),Dict("state"=>reduce(hcat,direct.solution.u),
            "mole_fractions"=>direct.mole_fractions,"pressure"=>direct.pressure,"distance"=>direct.solution.t))
        chain = surf_pfr_chain(path;output=joinpath(directory,"chain.native.csv"))
        npzwrite(joinpath(directory,"chain.native.npz"),Dict("mole_fractions"=>chain.mole_fractions,
            "mass_fractions"=>chain.mass_fractions,"coverages"=>chain.coverages,
            "pressure"=>chain.pressure,"distance"=>chain.distance,"residuals"=>chain.residuals))
    end
    reports = Dict{String,Any}()
    inlet_elements = flow.gas.ele_matrix*(flow.mass_fractions./flow.gas.MW)
    for case in ("flow","chain")
        ref = npzread(joinpath(directory,case=="flow" ? "flow.reference.npz" : "chain.refined.npz"))
        native = npzread(joinpath(directory,"$case.native.npz"))
        ng = flow.gas.n_species
        Y = case=="flow" ? native["state"][1:ng,:] : native["mass_fractions"]
        theta = case=="flow" ? native["state"][ng+1:end,:] : native["coverages"]
        elements = flow.gas.ele_matrix*(Y./flow.gas.MW)
        reference_elements = flow.gas.ele_matrix*(ref["mass_fractions"]./flow.gas.MW)
        reports[case] = Dict(
            "mole_fraction_absolute_error"=>maximum(abs,native["mole_fractions"]-ref["mole_fractions"]),
            "coverage_absolute_error"=>maximum(abs,theta-ref["coverages"]),
            "pressure_relative_error"=>maximum(abs.(native["pressure"]-ref["pressure"])./ref["pressure"]),
            "element_inventory_drift"=>maximum(abs,elements.-inlet_elements),
            "reference_element_inventory_drift"=>maximum(abs,reference_elements.-inlet_elements),
            "mass_flux_relative_drift"=>maximum(abs,vec(sum(Y;dims=1)).-1),
            "site_balance_drift"=>maximum(abs,vec(sum(theta;dims=1)).-1))
        @test native["distance"] ≈ ref["distance"] rtol=1e-14
        @test reports[case]["mole_fraction_absolute_error"] < 2e-7
        @test reports[case]["coverage_absolute_error"] < 2e-7
        @test reports[case]["pressure_relative_error"] < 1e-8
        @test reports[case]["element_inventory_drift"] < 1e-9
        @test reports[case]["mass_flux_relative_drift"] < 1e-9
        @test reports[case]["site_balance_drift"] < 1e-9
        @test minimum(Y) >= -1e-12
        @test minimum(theta) >= -1e-12
        if case == "flow"
            diagnostics = [surface_flow_diagnostics(flow,u) for u in eachcol(native["state"])]
            stationarity = maximum(d.stationary_relative_residual for d in diagnostics)
            momentum_error = maximum(abs(d.pressure+d.density*d.speed^2-flow.momentum_flux)/flow.momentum_flux for d in diagnostics)
            reports[case]["stationary_relative_residual"] = stationarity
            reports[case]["momentum_flux_relative_error"] = momentum_error
            @test stationarity < 1e-7
            @test momentum_error < 1e-13
            @test [d.density for d in diagnostics] ≈ ref["density"] rtol=1e-7
            @test [d.speed for d in diagnostics] ≈ ref["speed"] rtol=1e-7
            npzwrite(joinpath(directory,"flow.fluxes.npz"),Dict(
                "species_mass_flux"=>reduce(hcat,[d.species_mass_flux for d in diagnostics]),
                "elemental_flux"=>reduce(hcat,[d.elemental_flux for d in diagnostics])))
        else
            source = npzread(joinpath(directory,"chain.reference.npz"))
            reports[case]["source_default_mole_fraction_difference"] = maximum(abs,native["mole_fractions"]-source["mole_fractions"])
            reports[case]["source_default_element_inventory_drift"] = maximum(abs,
                flow.gas.ele_matrix*(source["mass_fractions"]./flow.gas.MW).-inlet_elements)
            @test maximum(native["residuals"]) < 1e-9
            npzwrite(joinpath(directory,"chain.fluxes.npz"),Dict(
                "species_mass_flux"=>flow.mass_flux.*Y,"elemental_flux"=>flow.mass_flux.*elements))
        end
        @show case reports[case]
    end
    open(joinpath(directory,"surface_flow_validation.toml"),"w") do io
        TOML.print(io,reports;sorted=true)
    end
    return reports
end

if abspath(PROGRAM_FILE)==@__FILE__
    length(ARGS)==1 || error("supply the prepared surface-flow reference directory")
    @testset "Cantera surface flow and 201-reactor chain" begin
        validate_surface_flows(ARGS[1])
    end
end

File: validation/surface_flow_cases.py
"""Exact source settings for Cantera surf_pfr_chain.py and surf_pfr.py.

The chain reports its first reactor outlet at z=0, as the source does. The
direct DAE reference is sampled at exact requested coordinates to avoid the
source CSV loop's pre-step coordinate/post-step state offset.
Both the unmodified chain tolerances and a tighter reference are retained,
because the source defaults accumulate measurable elemental drift over 201 cells.
https://cantera.org/dev/examples/python/reactors/surf_pfr_chain.html
https://cantera.org/dev/examples/python/reactors/surf_pfr.html
"""
from __future__ import annotations

import argparse
import importlib.util
import json
from pathlib import Path

import cantera as ct
import numpy as np


def main(directory):
    directory.mkdir(parents=True,exist_ok=True)
    spec=importlib.util.spec_from_file_location("export_surface",Path(__file__).resolve().parents[1]/"mechanism"/"export_surface.py")
    exporter=importlib.util.module_from_spec(spec);spec.loader.exec_module(exporter)
    exporter.export_surface("methane_pox_on_pt.yaml","Pt_surf",directory/"pox.surface.npz")
    def phase():
        s=ct.Interface("methane_pox_on_pt.yaml","Pt_surf")
        g=s.adjacent["gas"]
        s.TP=1073.15,ct.one_atm
        g.TPX=1073.15,ct.one_atm,"CH4:1,O2:1.5,AR:0.1"
        return s,g
    s,g=phase()
    A,porosity,length=1e-4,0.3,0.003
    mdot=0.4/60*g.density*A*porosity
    n=201;dz=length/(n-1)
    for filename,rtol,atol in (("chain.reference.npz",None,None),("chain.refined.npz",1e-11,1e-22)):
        s,g=phase()
        r=ct.IdealGasReactor(g,energy="off",volume=A*dz*porosity,clone=False)
        upstream=ct.Reservoir(g,clone=True);downstream=ct.Reservoir(g,clone=True)
        rs=ct.ReactorSurface(s,r,A=1e5*r.volume,clone=False)
        feed=ct.MassFlowController(upstream,r,mdot=mdot)
        outlet=ct.PressureController(r,downstream,primary=feed,K=1e-6)
        sim=ct.ReactorNet([r])
        if rtol is not None:sim.rtol,sim.atol=rtol,atol
        X,Y,theta,P=[],[],[],[]
        for k in range(n):
            upstream.phase.TDY=r.phase.TDY
            sim.reinitialize();sim.solve_steady()
            X.append(r.phase.X.copy());Y.append(r.phase.Y.copy());theta.append(rs.coverages.copy());P.append(r.phase.P)
        np.savez(directory/filename,distance=np.arange(n)*dz,mole_fractions=np.array(X).T,
                 mass_fractions=np.array(Y).T,coverages=np.array(theta).T,pressure=P,
                 mass_flow_rate=np.array([mdot]),rtol=np.array([sim.rtol]),atol=np.array([sim.atol]))
    s,g=phase()
    r=ct.FlowReactor(g,clone=False)
    r.area=A;r.mass_flow_rate=mdot;r.energy_enabled=False
    rs=ct.FlowReactorSurface(s,r,clone=False);rs.area=1e5*porosity*A
    sim=ct.ReactorNet([r]);sim.rtol=1e-10;sim.atol=1e-18;sim.max_steps=100000
    sim.initialize()
    distance=np.linspace(0,length,301)
    X,Y,theta,P,rho,speed=[],[],[],[],[],[]
    for z in distance:
        if z>0:sim.advance(z)
        X.append(r.phase.X.copy());Y.append(r.phase.Y.copy());theta.append(rs.coverages.copy())
        P.append(r.phase.P);rho.append(r.phase.density);speed.append(r.speed)
    np.savez(directory/"flow.reference.npz",distance=distance,mole_fractions=np.array(X).T,
             mass_fractions=np.array(Y).T,coverages=np.array(theta).T,pressure=P,density=rho,speed=speed,
             mass_flow_rate=np.array([mdot]))
    manifest={"cantera_version":ct.__version__,"mechanism":"methane_pox_on_pt.yaml",
        "temperature":1073.15,"pressure":ct.one_atm,"composition":"CH4:1,O2:1.5,AR:0.1",
        "length":length,"area":A,"porosity":porosity,"velocity":0.4/60,
        "catalyst_area_per_volume":1e5,"mass_flow_rate":mdot,"chain_reactors":n,
        "flow_sample_points":len(distance),"flow_rtol":sim.rtol,"flow_atol":sim.atol,
        "chain_refined_rtol":1e-11,"chain_refined_atol":1e-22}
    (directory/"manifest.json").write_text(json.dumps(manifest,indent=2)+"\n")


if __name__ == "__main__":
    parser=argparse.ArgumentParser(description=__doc__);parser.add_argument("directory",type=Path)
    main(parser.parse_args().directory)
